lib/custatevec/src/statevec.jl

```diff
@@ -1,16 +1,62 @@
+@inline with_cached_workspace(f::Base.Callable, size::Union{Integer,Function}, cached_ws::CuVector{UInt8}) =
+    with_cached_workspaces(f, UInt8, size, -1, cached_ws)
+
+@inline with_cached_workspace(f::Base.Callable, eltyp::Type{T}, size::Union{Integer,Function}, cached_ws::CuVector{T}) where {T} = with_cached_workspaces(f, T, size, -1, cached_ws)
+
+@inline with_cached_workspaces(f::Base.Callable, size_gpu::Union{Integer,Function}, size_cpu::Union{Integer,Function}, cached_ws::CuVector{UInt8}) = with_cached_workspaces(f, UInt8, size_gpu, size_cpu, cached_ws)
+
+function with_cached_workspaces(f::Base.Callable, eltyp::Type{T}, size_gpu::Union{Integer,Function}, size_cpu::Union{Integer,Function}, cached_ws::CuVector{T}) where {T}
+    get_size_gpu() = Int(isa(size_gpu, Integer) ? size_gpu : size_gpu()::Integer)
+    get_size_cpu() = Int(isa(size_cpu, Integer) ? size_cpu : size_cpu()::Integer)
+
+    sz_gpu = get_size_gpu() # this is bufferSize
+    sz_cpu = get_size_cpu()
+
+    workspace_gpu = nothing 
+    if sz_gpu <= length(cached_ws)
+        workspace_gpu = cached_ws
+    else
+        try
+            while workspace_gpu === nothing || length(workspace_gpu) < sz_gpu
+                workspace_gpu = CuVector{T}(undef, sz_gpu)
+                sz_gpu = get_size_gpu()
+            end
+        catch ex
+            isa(ex, OutOfGPUMemoryError) || rethrow()
+            workspace_gpu = cached_ws
+        end
+    end
+    workspace_gpu = workspace_gpu::CuVector{T}
+
+    # use & free
+    try
+        # size_cpu == -1 means that we don't need a CPU workspace
+        if sz_cpu == -1
+            f(workspace_gpu)
+        else
+            workspace_cpu = Vector{T}(undef, sz_cpu)
+            f(workspace_gpu, workspace_cpu)
+        end
+    finally
+        cached_ws = workspace_gpu
+    end
+end
+
 function applyPauliExp!(sv::CuStateVec, theta::Float64, paulis::Vector{<:Pauli}, targets::Vector{Int32}, controls::Vector{Int32}, controlValues::Vector{Int32}=fill(one(Int32), length(controls)))
     cupaulis = CuStateVecPauli.(paulis)
-    custatevecApplyPauliRotation(handle(), sv.data, eltype(sv), sv.nbits, theta, cupaulis, targets, length(targets), controls, controlValues, length(controls))
+    custatevecApplyPauliRotation(handle().handle, sv.data, eltype(sv), sv.nbits, theta, cupaulis, targets, length(targets), controls, controlValues, length(controls))
```

Comment:
You shouldn't have to do the `handle().handle` because of the remaining `unsafe_convert` method.